### PR TITLE
[UNDERTOW-2287] Add ACL style ip-match predicate

### PR DESCRIPTION
**Pull request by baranowb**
Issue: https://issues.redhat.com/browse/UNDERTOW-2287

### CODE REVIEW

**Review by fl4via — APPROVED** [review]

File: core/src/main/java/io/undertow/predicate/ip/IPMatchBase.java [hunk 1]
```diff
@@ -0,0 +1,388 @@
...
+        }
+    }
+
+    private void addIpV4ExactMatch(final String peer, final boolean deny) {
+        String[] parts = peer.split("\\.");
+        byte[] bytes = {(byte) Integer.parseInt(parts[0]), (byte) Integer.parseInt(parts[1]), (byte) Integer.parseInt(parts[2]), (byte) Integer.parseInt(parts[3])};
+        ipv4Matches.add(new ExactIpV4PeerMatch(peer, bytes, deny));
+    }
+
+    protected abstract class PeerMatch {
```
Comment (ropalka) [inline core/src/main/java/io/undertow/predicate/ip/IPMatchBase.java:278]: static modifier got lost during the refactoring

File: core/src/main/java/io/undertow/predicate/ip/IPMatchBase.java [hunk 2]
```diff
@@ -0,0 +1,388 @@
...
+                    + ", pattern='" + pattern + '\''
+                    + '}';
+        }
+
+        public String toPredicateString() {
+            return pattern + " " + (deny ? "deny" : "allow");
+        }
+    }
+
+    protected class ExactIpV4PeerMatch extends PeerMatch {
```
Comment (ropalka) [inline core/src/main/java/io/undertow/predicate/ip/IPMatchBase.java:306]: Same here, static modifier got lost during the refactoring

File: core/src/main/java/io/undertow/predicate/ip/IPMatchBase.java [hunk 3]
```diff
@@ -0,0 +1,388 @@
...
+            this.address = address;
+        }
+
+        @Override
+        public boolean matches(final InetAddress address) {
+            return Arrays.equals(address.getAddress(), this.address);
+        }
+    }
+
+    protected class ExactIpV6PeerMatch extends PeerMatch {
```
Comment (ropalka) [inline core/src/main/java/io/undertow/predicate/ip/IPMatchBase.java:321]: Same here, static modifier got lost during the refactoring

File: core/src/main/java/io/undertow/predicate/ip/IPMatchBase.java [hunk 4]
```diff
@@ -0,0 +1,388 @@
...
+            this.address = address;
+        }
+
+        @Override
+        public boolean matches(final InetAddress address) {
+            return Arrays.equals(address.getAddress(), this.address);
+        }
+    }
+
+    protected class PrefixIpV4PeerMatch extends PeerMatch {
```
Comment (ropalka) [inline core/src/main/java/io/undertow/predicate/ip/IPMatchBase.java:336]: Same here, static modifier got lost during the refactoring

File: core/src/main/java/io/undertow/predicate/ip/IPMatchBase.java [hunk 5]
```diff
@@ -0,0 +1,388 @@
...
+            byte[] bytes = address.getAddress();
+            if (bytes == null) {
+                return false;
+            }
+            int addressInt = ((bytes[0] & 0xFF) << 24) | ((bytes[1] & 0xFF) << 16) | ((bytes[2] & 0xFF) << 8) | (bytes[3] & 0xFF);
+            return (addressInt & mask) == prefix;
+        }
+    }
+
+    protected class PrefixIpV6PeerMatch extends PeerMatch {
```
Comment (ropalka) [inline core/src/main/java/io/undertow/predicate/ip/IPMatchBase.java:358]: Same here, static modifier got lost during the refactoring

**Review by ropalka — CHANGES_REQUESTED** [review]
Please add back the static class modifier otherwise every subclass of IPMatchBase will have its own inner classes created and it will cost much more memory.

**Review by ropalka — APPROVED** [review]

**Pull request merged**